### ISSUE TITLE
.NET 8.0.10 breaks MAUI builds

### ISSUE THREAD

**Issue by timjamwoo** [dotnet/maui#21170]
### Description

I've updated our MAUI project to 8.0.10 today and cannot build for Android or iOS. I updated by changing the `<MauiVersion>` tag in our .csproj to `<MauiVersion>8.0.10</MauiVersion>`.

I see this warning in the build log:
> 1>CSC: Warning CS8785 : Generator 'CodeBehindGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentNullException' with message 'Value cannot be null. (Parameter 'second')'.

This is followed by many errors related to XAML files not correctly linking to their corresponding codebehind (e.g. `Error CS0103 : The name 'InitializeComponent' does not exist in the current context`).

I have tried this on the latest versions of VS 2022 for Windows, and Rider for Mac, targeting Android and iOS respectively. 

### Steps to Reproduce

1. Ensure you have latest Visual Studio/.NET installed
2. Run `dotnet workload update`
3. In an existing MAUI project, add/update the MauiVersion tag to the csproj: <MauiVersion>8.0.10</MauiVersion>
4. Try to build your project

### Version with bug

8.0.10

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.7

### Affected platforms

iOS, Android

### Affected platform versions

Android 10, iOS 17.3.1

### Did you find any workaround?

Rollback to .NET 8.0.7.

### Relevant log output

1>CSC: Warning CS8785 : Generator 'CodeBehindGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentNullException' with message 'Value cannot be null. (Parameter 'second')'.

**Comment by developer9969** [comment]
Just to say that I have run our azure devops pipeline against 8.0.10 and all worked building both iOS and android

**Comment by Domik234** [comment]
Is this happening on application or library?

I am asking because many ppl use MSBuild.Sdk.Extras to release Maui/Xamarin library but there is propably not support for newer .NET versions so MauiXaml and similar may be problematic.

---
Btw: There is also possible to emit generated files (purposely) by using @ .csproj:
`<EmitCompilerGeneratedFiles>false</EmitCompilerGeneratedFiles>`

**Comment by timjamwoo** [comment]
@Domik234 this is a MAUI application.

**Comment by MAUIoxo** [comment]
I am not sure, is there an official`8.0.10` version or did you want to take a nightly build which would be something like `8.0.20-nightly.10311`?

**Comment by timjamwoo** [comment]
> I am not sure, is there an official`8.0.10` version or did you want to take a nightly build which would be something like `8.0.20-nightly.10311`?

@MAUIoxo there is an [official 8.0.10 release](https://github.com/dotnet/maui/releases/tag/8.0.10).

**Comment by Domik234** [comment]
> @Domik234 this is a MAUI application.

Yeah, this is "maui" repo. But I've asked the question because many people tries to support both using one nuget (I am also using this method and works for me) and I know that there is this specific problem. 😁

---

Try this `https://github.com/Domik234/IssueMaui-21170` (MauiApp2/MauiApp2.sln)
It contains only clear project + library with one Xaml ContentView.
Maui Version: 8.0.10
Platfoms: Android, iOS

This works for me as it should so cannot repro the problem.

Also don't forget to remove "bin" and "obj" folder after the version change at your project.

---

> I am not sure, is there an official`8.0.10` version or did you want to take a nightly build which would be something like `8.0.20-nightly.10311`?

There was released Maui Version 8.0.10 yesterday. (SR3?) 😀

**Comment by timjamwoo** [comment]
> Try this `https://github.com/Domik234/IssueMaui-21170` (MauiApp2/MauiApp2.sln) It contains only clear project + library with one Xaml ContentView. Maui Version: 8.0.10 Platfoms: Android, iOS
> 
> This works for me as it should so cannot repro the problem.

Thanks @Domik234 - I see that your test project doesn't include the `<MauiVersion>8.0.10</MauiVersion>` tag in the csproj - I was under the impression that for our 8.0.7 migration this was necessary to specify which release we're targeting. Am I mistaken?

**Comment by Axemasta** [comment]
I'm seeing the same issue for a maui app thats been in development for some time, the build is fine on `8.0.7` and if I only change that version to `8.0.10`, I see the following error in the build output:
```
7>CSC: Warning CS8785 : Generator 'CodeBehindGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentNullException' with message 'Value cannot be null. (Parameter 'second')'.
```

And suddenly I'm greeted with hundreds of errors, all related to `InitializeComponent` not being found, this would make sense is the code behind generator fell over.

I've confirmed I'm using the latest 8.0.10 package, I've cleared my bin/obj folder and restored the solution and the result is the same.

I've tested this some of my demo maui apps that don't use any codebehind features (except `InitializeComponent`) and they do build. I see in the release notes for 8.0.10 that there are changes to the codebehind generator (#20524), could this be the cause of these new failures?

This looks like quite a serious issue that will brick 8.0.10 for any existing maui users, I'm happy to provide more verbose logs for investigation from my failing project.

**Comment by Axemasta** [comment]
I turned on detailed logging and have a stack trace for the argument nulle xception thrown by the code gen:
```
3>CSC: Warning CS8785 : Generator 'CodeBehindGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentNullException' with message 'Value cannot be null. (Parameter 'second')'.
System.ArgumentNullException: Value cannot be null. (Parameter 'second')
at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
at System.Linq.Enumerable.SequenceEqual[TSource](IEnumerable`1 first, IEnumerable`1 second, IEqualityComparer`1 comparer)
at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.GetTypeName(XmlType xmlType, Compilation compilation, AssemblyCaches xmlnsCache, IDictionary`2 typeCache)
at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.TryParseXaml(XamlProjectItem parseResult, String uid, Compilation compilation, AssemblyCaches xmlnsCache, IDictionary`2 typeCache, CancellationToken cancellationToken, String& accessModifier, String& rootType, String& rootClrNamespace, Boolean& generateDefaultCtor, Boolean& addXamlCompilationAttribute, Boolean& hideFromIntellisense, Boolean& xamlResourceIdOnly, String& baseType, IEnumerable`1& namedFields)
at Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator.GenerateXamlCodeBehind(XamlProjectItem xamlItem, Compilation compilation, SourceProductionContext context, AssemblyCaches xmlnsCache, IDictionary`2 typeCache)
at Microsoft.CodeAnalysis.UserFunctionExtensions.<>c__DisplayClass3_0`2.<WrapUserAction>b__0(TInput1 input1, TInput2 input2, CancellationToken token)
```

**Comment by jonathanpeppers** [comment]
Probably some null check is needed here: 
* https://github.com/dotnet/maui/commit/4fa29bc4366b70a40fb91f8fe9dc90474e49edeb#diff-f6b72a5f642bbafa89a856d75dddbe9a9f3d7a86ec66fe1069ec0738ced02d02R782

`$(Nullable)=enable` is set in this project, but I guess it didn't catch this.

/cc @@mgoertz-msft

**Comment by mgoertz-msft** [comment]
I'll take a look...

**Comment by mrwcjoughinaaf** [comment]
I am getting the same error on a big Xamarin to Maui conversion project I am working on. Have had to move the packages back to 8.0.7 as none of the workarounds mentioned work.

**Comment by mgoertz-msft** [comment]
I have a fix. See associated PR #21192

**Comment by ryanalford-suzy** [comment]
Hopefully a fix can be pushed out quickly. 8.0.10 fixes a number of issues we are running into, but now causes hundreds of build errors.

**Comment by mrwcjoughinaaf** [comment]
> I have a fix. See associated PR #21192

Thanks @mgoertz-msft . I see there are tests failing that are preventing the PR from being allowed to be approved - please can you resolve those asap as we really need the fix - I have to explain to the client today that I wasted a day on this issue and would like to be able to say that the new version of the package will be released very soon.

**Comment by Dave-Quick** [comment]
I too have run into this.

**Comment by ryanalford-suzy** [comment]
@mgoertz-msft Do you happen to have an idea of when this fix will be released? Looks like it was merged into the "SR3" branch. Just wondering if there was a timeline on it's release.

**Comment by PureWeen** [comment]
Can you test with the latest nightly build? 
https://github.com/dotnet/maui/wiki/Nightly-Builds

**Comment by PureWeen** [comment]
We're prepping a patch release to SR3 now
It'd be helpful if you could test nightly to validate it fixes your issue

**Comment by ryanalford-suzy** [comment]
Using `8.0.20-nightly.10336`, I do not get build errors anymore. Looks like it's fixed!!! @PureWeen

**Comment by danardelean** [comment]
I also tested with 8.0.14 (from the nightly stream) and the build works

**Comment by jasho** [comment]
We also ran into this issue and it's currently blocking our release. Tested the 8.0.20-nightly.10336 and can confirm that the build works fine.

**Comment by Redth** [comment]
8.0.14 was released to NuGet which should fix this

**Comment by Axemasta** [comment]
Thanks for the quick fix, confirmed working on my app using `8.0.14` ❤️
